### ISSUE TITLE
Even more spaceship quasar problems

### ISSUE THREAD

**Issue by i30817** [adventuregamestudio/ags#1117]
**Describe the bug**

left clicking the TV on the first room crashes ags, it depends on the number of clicks (some images in particular seem to crash). 
```
An internal error has occurred. Please note down the following information.
If the problem persists, post the details on the AGS Technical Forum.
(ACI version 3.5.0.26)

Error: There was an error drawing object 3. Its current sprite, 178, is invalid.

***** ENGINE HAS SHUTDOWN

```
I think this is just a problem of the game, I went a bit far back in the engine and it always seemed to happen.

Moving two rooms south once and through the other door once crashes the game unavoidably (very early after the bug started even the first room south).
```
An error has occurred. Please contact the game author for support, as this is likely to be a scripting error and not a bug in AGS.
(ACI version 3.5.0.26)

in "room4.asc", line 155

Error: SetObjectFrame: invalid frame number used
```
This is both a problem of the game version i have and a regression. It stops happening before df01670afdcc08adbcedbaf42671da6b8f9821fb, which changed the `fra(me) <0` case from 'keep it zero in the struct instead of copying' to 'make it a it a error'. There is a effect on this room of some electricity bolts which look like the culprit for having a 'frame' < 0.

I find it strange that it appears to happen in the first room right after that commit, and now in master only in the second room.

**AGS Version**
Adventure Game Studio v3.5 Interpreter
Copyright (c) 1999-2011 Chris Jones and 2011-2020 others
ACI version 3.5.0.26

self-built after the fix for #1116 with `make clean --directory=Engine; make USE_BUILT_IN_LIBSRC=1 --directory=Engine`

**Game**
Spaceship quasar

**To Reproduce**
1. Walk south then through the other door or click several times on the tv

**Comment by i30817** [comment]
This might also be a uninitialized variable in the optional arguments defaulting to something not 0 from UB, not necessarily a error of the commit.

**Comment by rofl0r** [comment]
anyone got the 2011 freeware version of the game around? couldn't find it in agsarchives.com torrent

**Comment by i30817** [comment]
Interesting. They're pretty different games but the screen where it crashes looks almost equal (down the first screen, through the door).

And i realize now that there is something very very wrong with the TV in the 2.0 version of the game i have (besides crashing). Namely the images it displays are not contained in the TV like in that version, they're on a square on the top right.

**Comment by rofl0r** [comment]
freeware version doesn't crash here. i'm using c38b937273f21ec6002cde1dec9b1bead5f7833c with a couple custom modifications on top.

**Comment by i30817** [comment]
Yeah, because it 'stops' crashing before https://github.com/adventuregamestudio/ags/commit/df01670afdcc08adbcedbaf42671da6b8f9821fb and yours is almost a year before?

**Comment by rofl0r** [comment]
oh, didn't see that. is the door you're talking about the one of "engine room" ?

**Comment by i30817** [comment]
Yes, that's the screen where it crashes. I think it's the lightning animation on top of the half cylinder on that screen.

**Comment by rofl0r** [comment]
my money is on that the changes to SetObjectFrame() broke it

**Comment by i30817** [comment]
Yeah, the code i put in above avoids the crash. A interesting detail is that right at https://github.com/adventuregamestudio/ags/commit/df01670afdcc08adbcedbaf42671da6b8f9821fb the game (at least the deluxe version haven't tried with the demo) crashed on a different screen, in the cargo bay (which is just 1 room south of the start).

**Comment by ivan-mogilko** [comment]
For the reference, I have this game with a bundle on Steam: https://store.steampowered.com/app/551840/Crystal_Shard_Adventure_Bundle/

**Comment by ivan-mogilko** [comment]
To be frank, the changes to df01670 were made without much considering of legacy games at that time, because their rules are unclear. I might need to adjust SetObjectFrame once more to account for this.

**Comment by rofl0r** [comment]
> I might need to adjust SetObjectFrame once more to account for this.

how about just doing 

`if(gamever < 351) { old code ; } else { new code; }`

**Comment by rofl0r** [comment]
btw the ifdef in the .sh agscc header looks like it would only expose the new behaviour in that exact version.
is there no way to do something like `#if SCRIPT_API < 351 ... #else` ?

**Comment by i30817** [comment]
Aren't ifdefs pointless to run the games with a single executable (like is typically done in linux)? I don't think there is a point in restricting old legacy ags to 'can only run versions of the executable it compiled/came with'

**Comment by ivan-mogilko** [comment]
> > I might need to adjust SetObjectFrame once more to account for this.
> 
> how about just doing
> 
> `if(gamever < 351) { old code ; } else { new code; }`

Of course, but the question was what is the old code, because it already could have passed more than one iteration since.

**Comment by ivan-mogilko** [comment]
The irritating thing here is that the wrong frame error occurs in case of a direct call to SetObjectFrame(), not SetView(), which originally could alter loop and frame. This means that these parameters had to be passed explicitly by game authors, and the reason it did not crash in previous version was because SetObjectFrame did that unexpected silent fixup to put an argument in the range.

**Comment by rofl0r** [comment]
> @i30817 I pushed the fix, which is practically what you suggested above.

link: e82746e78248635b6c338846aba36b2fa6e036ae

**Comment by i30817** [comment]
> @i30817 I pushed the fix, which is practically what you suggested above.
> 
> I was not able to reproduce the TV crash though. How long do I have to click on the screen?

It doesn't happen in the demo that rofl0r linked above but not too long. It seems to happen immediately at the start, at a latter save of the game when i had another npc in the room i couldn't do it (i saw another image with a different size which seemed like a small green cross on a black background too).
It's possible the game was updated meanwhile and this bug no longer happens though, so i'm not too worried about it.

Here is a quick crash at the start:
![Peek 2020-09-09 00-51](https://user-images.githubusercontent.com/757900/92538588-ffae7300-f236-11ea-8974-19fe16dfb61b.gif)

**Comment by ivan-mogilko** [comment]
> It doesn't happen in the demo that rofl0r linked above but not too long. 

Could you rephrase this? it does or does not happen in the linked demo?

**Comment by i30817** [comment]
It does not happen in the demo, it happens in the version of the game 'quasar deluxe edition v2.0' (from the main menu text) i have here.

**Comment by i30817** [comment]
Btw, the fix fixed the other crash of course.

**Comment by i30817** [comment]
I did the obvious thing and tried the game in wine like i should have: the crash exist there, so it's from the game data. Please close, it's fixed as much as it will be.

**Comment by ivan-mogilko** [comment]
Alright.
